crates/starknet-os/src/hints/kzg.rs

```diff
@@ -0,0 +1,269 @@
+use cairo_vm::types::relocatable::MaybeRelocatable;
+use cairo_vm::vm::errors::hint_errors::HintError;
+use cairo_vm::vm::vm_core::VirtualMachine;
+use cairo_vm::Felt252;
+use indoc::indoc;
+use num_bigint::BigInt;
+use num_traits::{Num, One, Zero};
+
+const FIELD_ELEMENTS_PER_BLOB: usize = 4096;
+const COMMITMENT_BYTES_LENGTH: usize = 48;
+const COMMITMENT_LOW_BIT_LENGTH: usize = (COMMITMENT_BYTES_LENGTH * 8) / 2;
+const BLOB_SUBGROUP_GENERATOR: &str = "39033254847818212395286706435128746857159659164139250548781411570340225835782";
+const BLS_PRIME: &str = "52435875175126190479447740508185965837690552500527637822603658699938581184513";
+
+fn fft(coeffs: &[BigInt], generator: &BigInt, prime: &BigInt, bit_reversed: bool) -> Result<Vec<BigInt>, io::Error> {
+    fn _fft(coeffs: &[BigInt], group: &[BigInt], prime: &BigInt) -> Vec<BigInt> {
+        if coeffs.len() == 1 {
+            return coeffs.to_vec();
+        }
+
+        let f_even = _fft(
+            &coeffs.iter().step_by(2).cloned().collect::<Vec<_>>(),
+            &group.iter().step_by(2).cloned().collect::<Vec<_>>(),
+            prime,
+        );
+        let f_odd = _fft(
+            &coeffs.iter().skip(1).step_by(2).cloned().collect::<Vec<_>>(),
+            &group.iter().step_by(2).cloned().collect::<Vec<_>>(),
+            prime,
+        );
+
+        let group_mul_f_odd: Vec<BigInt> =
+            group.iter().take(f_odd.len()).zip(f_odd.iter()).map(|(g, f)| (g * f) % prime).collect();
+
+        let mut result = Vec::with_capacity(coeffs.len());
+        for i in 0..f_even.len() {
+            result.push((f_even[i].clone() + &group_mul_f_odd[i]) % prime);
+        }
+        for i in 0..f_even.len() {
+            // Ensure non-negative diff
+            let diff = (f_even[i].clone() - &group_mul_f_odd[i] + prime) % prime;
+            result.push(diff);
+        }
+
+        result
+    }
+
+    if coeffs.is_empty() {
+        return Ok(vec![]);
+    }
+
+    let coeffs_len = coeffs.len();
+    assert!(coeffs_len.is_power_of_two(), "Length is not a power of two.");
+
+    let mut group = vec![BigInt::one()];
+    for _ in 1..coeffs_len {
+        let last = group.last().ok_or(io::Error::other("Group is missing last element"))?;
+        group.push((last * generator) % prime);
+    }
+
+    let mut values = _fft(coeffs, &group, prime);
+
+    if bit_reversed {
+        let width = coeffs_len.trailing_zeros() as usize;
+        let perm = (0..coeffs_len)
+            .map(|i| {
+                let binary = format!("{:0width$b}", i, width = width);
+                usize::from_str_radix(&binary.chars().rev().collect::<String>(), 2)
+                    .map_err(|_| io::Error::other("Invalid binary cast to usize"))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        values = perm.into_iter().map(|i| values[i].clone()).collect();
+    }
+
+    Ok(values)
+}
+
+fn split_commitment(num: BigInt) -> (BigInt, BigInt) {
```

Comment:
I don't see this code in `cairo-lang`, is this closed source? Please point me to the source code for reference.